SEP/SEP-0005/sep-0005.md

```diff
@@ -0,0 +1,170 @@
+## PREFER, a new way to select preferred answers
+
+* Short Name: prefer
+* SEP Number: [SEP-0005](sep-0005.md)
+* Author: [Jerven Bolleman](https://github.com/jervenbolleman)
+
+## Authors
+
+Jerven Bolleman
+
+## Abstract
+
+RDF and SPARQL have excellent support for multiple languages and this is widely used. However, there is a lack of selecting a preferred language given a set of possibilities. 
+
+## Motivation
+
+While technically possible in SPARQL 1.1. We would like to have an easier to write query syntax that is potentially easier to optimize for query engines. Even if the final syntax can be rewritten to SPARQL 1.1 equivalent constructs.
+
+
+
+Given the requirement to select labels/properties in languages in an order of user preference we currently need to use the lang() and langmatches() functions either in a series of optional blocks with tests for boundness, or a subselect with an ordering.
+
+
+```sparql
+SELECT 
+  ?thing 
+  ?label
+WHERE {
+  OPTIONAL {
+    ?thing rdfs:label ?label .
+    FILTER(langMatches(lang(?label), “fr”)
+  }
+  OPTIONAL {
+    FILTER(! BOUND(?label))
+    ?thing rdfs:label ?label .
+    FILTER(langMatches(lang(?label), “en”)
+  }
+  FILTER(BOUND(?label) && BOUND(?thing))
+}
+```
+Gives an example of the use with the optional solution.
+```sparql
+SELECT
+  ?thing
+  ?label
+WHERE
+{ { 
+ SELECT  
+   ?thing 
+   ?label 
+   ?order
+ WHERE
+ { VALUES ( ?lang ?order ) {
+             ( "fr" 1 )
+             ( "en" 2 )}
+   ?thing  rdfs:label  ?label
+   FILTER langMatches(lang(?label), ?lang)
+ }
+
+ GROUP BY ?thing ?label ?order
+ ORDER BY ?order
+ LIMIT   1
+ }
+ }
+```
+
+
+Neither variety are easy to write, and neither can take into account preferential settings regarding languages from a browser or http header.
+
+
+
+The proposal is to introduce a new keyword `PREFER` and functionality in the syntax and abstract algebra.
+
+
+
+The same query as above would look like this in the proposed syntax.
+
+
+```
+SELECT ?label 
+
+WHERE
+
+{
+
+  [] rdfs:label PREFER(?label FROM langMatches(lang(?label), "FR"),
+
+                         langMatches(lang(?label), "EN"), 
+
+                         langMatches(lang(?label), "NL")) .
+
+}
+```
+
+We will also introduce a second new keyword `ACCEPTLANG`, to refer to taking the order from the accept-lang http header provided by the client and a new shorthand for testing if a value has a matching language string `hasLang(?label, "EN")`.
```

Comment:
This seems problematic to me since it bleeds details of the HTTP protocol down into the query language.  What would be the meaning of this in a context where no HTTP is involved?

e.g. querying an entirely local dataset via a command line interface

Any specification would need to ensure that query engines are free to ignore this in contexts where it does not make sense

Reply:
I agree that this is problematic. It's probably a good idea to think about how Query works around a similar separation for [specifying an RDF Dataset](https://www.w3.org/TR/sparql11-query/#specifyingDataset). In that case, a dataset can be service-defined, specified in the query, or specified in the protocol.

Reply:
Very good point! for CLI it might look at the locale environment setting at least on linux. I suspect that windows will have an equivalent setting.